File: container-documentapi/src/vespa/documentapi/messagebus/replymerger.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "replymerger.h"
#include "documentprotocol.h"
#include <vespa/documentapi/messagebus/messages/removedocumentreply.h>
#include <vespa/documentapi/messagebus/messages/updatedocumentreply.h>
#include <vespa/documentapi/messagebus/messages/getdocumentreply.h>
#include <vespa/messagebus/emptyreply.h>
#include <cassert>

namespace documentapi {

ReplyMerger::ReplyMerger()
    : _error(),
      _ignored(),
      _successReply(0),
      _successIndex(0)
{
}

ReplyMerger::~ReplyMerger() {}

ReplyMerger::Result::Result(uint32_t successIdx,
                            std::unique_ptr<mbus::Reply> generatedReply)
    : _generatedReply(std::move(generatedReply)),
      _successIdx(successIdx)
{
}

ReplyMerger::Result::Result(Result&& o)
    : _generatedReply(std::move(o._generatedReply)),
      _successIdx(o._successIdx)
{
}

bool
ReplyMerger::Result::hasGeneratedReply() const
{
    return (_generatedReply.get() != 0);
}

bool
ReplyMerger::Result::isSuccessful() const
{
    return !hasGeneratedReply();
}

std::unique_ptr<mbus::Reply>
ReplyMerger::Result::releaseGeneratedReply()
{
    assert(hasGeneratedReply());
    return std::move(_generatedReply);
}

uint32_t
ReplyMerger::Result::getSuccessfulReplyIndex()
{
    assert(!hasGeneratedReply());
    return _successIdx;
}

void
ReplyMerger::merge(uint32_t idx, const mbus::Reply& r)
{
    if (r.hasErrors()) {
        mergeAllReplyErrors(r);
    } else {
        updateStateWithSuccessfulReply(idx, r);
    }
}

bool
ReplyMerger::resourceWasFound(const mbus::Reply& r) const
{
    switch (r.getType()) {
    case DocumentProtocol::REPLY_REMOVEDOCUMENT:
        return static_cast<const RemoveDocumentReply&>(r).wasFound();
    case DocumentProtocol::REPLY_UPDATEDOCUMENT:
        return static_cast<const UpdateDocumentReply&>(r).wasFound();
    case DocumentProtocol::REPLY_GETDOCUMENT:
        return static_cast<const GetDocumentReply&>(r).getLastModified() != 0;
    default:
        return false;
    }
}

// Precondition: _successReply != 0
bool
ReplyMerger::replyIsBetterThanCurrent(const mbus::Reply& r) const
{
    return resourceWasFound(r) && !resourceWasFound(*_successReply);
}

void
ReplyMerger::setCurrentBestReply(uint32_t idx, const mbus::Reply& r)
{
    _successIndex = idx;
    _successReply = &r;
}

void
ReplyMerger::updateStateWithSuccessfulReply(uint32_t idx, const mbus::Reply& r)
{
    if (!_successReply || replyIsBetterThanCurrent(r)) {
        setCurrentBestReply(idx, r);
    }
}

void
ReplyMerger::mergeAllReplyErrors(const mbus::Reply& r)
{
    if (handleReplyWithOnlyIgnoredErrors(r)) {
        return;
    }
    if (!_error.get()) {
        _error.reset(new mbus::EmptyReply());
    }
    for (uint32_t i = 0; i < r.getNumErrors(); ++i) {
        _error->addError(r.getError(i));
    }
}

bool
ReplyMerger::handleReplyWithOnlyIgnoredErrors(const mbus::Reply& r)
{
    if (DocumentProtocol::hasOnlyErrorsOfType(r, DocumentProtocol::ERROR_MESSAGE_IGNORED)) {
        if (!_ignored.get()) {
            _ignored.reset(new mbus::EmptyReply());
        }
        _ignored->addError(r.getError(0));
        return true;
    }
    return false;
}

bool
ReplyMerger::shouldReturnErrorReply() const
{
    if (_error.get()) {
        return true;
    }
    return (_ignored.get() && !_successReply);
}

std::unique_ptr<mbus::Reply>
ReplyMerger::releaseGeneratedErrorReply()
{
    if (_error.get()) {
        return std::move(_error);
    } else {
        assert(_ignored.get());
        return std::move(_ignored);
    }
}

bool
ReplyMerger::successfullyMergedAtLeastOneReply() const
{
    return (_successReply != 0);
}

ReplyMerger::Result
ReplyMerger::createEmptyReplyResult() const
{
    return Result(0u, std::unique_ptr<mbus::Reply>(new mbus::EmptyReply()));
}

ReplyMerger::Result
ReplyMerger::mergedReply()
{
    std::unique_ptr<mbus::Reply> generated;
    if (shouldReturnErrorReply()) {
        generated = releaseGeneratedErrorReply();
    } else if (!successfullyMergedAtLeastOneReply()) {
        return createEmptyReplyResult();
    }
    return Result(_successIndex, std::move(generated));
}

} // documentapi

File: documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/ReplyMerger.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.documentapi.messagebus.protocol;

import com.yahoo.collections.Tuple2;
import com.yahoo.messagebus.EmptyReply;
import com.yahoo.messagebus.Error;
import com.yahoo.messagebus.Reply;

/**
 * Encapsulated logic for merging replies from 1-n related DocumentProtocol messages.
 * For internal use only. Not multithread safe.
 */
final class ReplyMerger {

    private Reply successReply = null;
    private int successIndex = -1;
    private Reply error = null;
    private Reply ignore = null;

    public void merge(int i, Reply r) {
        if (r.hasErrors()) {
            mergeAllReplyErrors(r);
        } else {
            updateStateWithSuccessfulReply(i, r);
        }
    }

    private boolean resourceWasFound(Reply r) {
        if (r instanceof RemoveDocumentReply) {
            return ((RemoveDocumentReply) r).wasFound();
        }
        if (r instanceof UpdateDocumentReply) {
            return ((UpdateDocumentReply) r).wasFound();
        }
        if (r instanceof GetDocumentReply) {
            return ((GetDocumentReply) r).getLastModified() > 0;
        }
        return false;
    }

    private boolean replyIsBetterThanCurrent(Reply r) {
        return resourceWasFound(r) && !resourceWasFound(successReply);
    }

    private void updateStateWithSuccessfulReply(int i, Reply r) {
        if (successReply == null || replyIsBetterThanCurrent(r)) {
            setCurrentBestReply(i, r);
        }
    }

    private void setCurrentBestReply(int i, Reply r) {
        successReply = r;
        successIndex = i;
    }

    private void mergeAllReplyErrors(Reply r) {
        if (handleReplyWithOnlyIgnoredErrors(r)) {
            return;
        }
        if (error == null) {
            error = r;
        }
        else if (mostSevereErrorCode(r) > mostSevereErrorCode(error)) {
            error.getErrors().forEach(r::addError);
            error = r;
        }
        else {
            r.getErrors().forEach(error::addError);
        }
    }

    private static int mostSevereErrorCode(Reply reply) {
        return reply.getErrors().mapToInt(Error::getCode).max()
                .orElseThrow(() -> new IllegalArgumentException(reply + " has no errors"));
    }

    private boolean handleReplyWithOnlyIgnoredErrors(Reply r) {
        if (DocumentProtocol.hasOnlyErrorsOfType(r, DocumentProtocol.ERROR_MESSAGE_IGNORED)) {
            if (ignore == null) {
                ignore = new EmptyReply();
            }
            ignore.addError(r.getError(0));
            return true;
        }
        return false;
    }

    private boolean shouldReturnErrorReply() {
        return (error != null || (ignore != null && successReply == null));
    }

    private Tuple2<Integer, Reply> createMergedErrorReplyResult() {
        if (error != null) {
            return new Tuple2<>(null, error);
        }
        if (ignore != null && successReply == null) {
            return new Tuple2<>(null, ignore);
        }
        throw new IllegalStateException("createMergedErrorReplyResult called without error");
    }

    private boolean successfullyMergedAtLeastOneReply() {
        return successReply != null;
    }

    private Tuple2<Integer, Reply> createEmptyReplyResult() {
        return new Tuple2<>(null, new EmptyReply());
    }

    public Tuple2<Integer, Reply> mergedReply() {
        if (shouldReturnErrorReply()) {
            return createMergedErrorReplyResult();
        } else if (!successfullyMergedAtLeastOneReply()) {
            return createEmptyReplyResult();
        }
        return new Tuple2<>(successIndex, successReply);
    }

}
